README.md
[<img src='/wmap_logo.jpg' width='350' height='350'>](https://github.com/yangsec888/wmap)
=====================

- [What's this program for?](#whats-this-program-for)
- [WMAP in Motion](#wmap-in-motion)
- [Program Version](#program-version)
- [Author Contact](#author-contact)
- [Installation](#installation)
- [Before Using This Program](#before-using-this-program)
- [More Document(s)](#more-documents)
- [Bug Report or Feature Request](#bug-report-or-feature-request)
- [Legal Disclaimer](#legal-disclaimer)

# OWASP Web Mapper README


## What's this program for?
This program is designed for the web application asset discovery and tracking. It was originally developed to
to cover the gaps of a similar commercial product. Over the time it grows to be a more capable and complete replacement (IMHO).


## WMAP in Motion
Use the demo web app build on top of wmap gem: http://wmap.io


## Program Version
The latest release is version 2.4.5. as of fall 2018. Please refer to the CHANGELOG.md for the program's history information.


## Author Contact
This program is designed and developed by Yang Li. You can reach him by Email: <yang.li@owasp.org>


## Installation
To take full power of this program, you would need an *nix flavor machine with direct Internet access. I have installed it successfully on both Mac and Linux machines. You'll also need the Ruby environment being setup properly. The easiest way to install OWASP Web Mapper is by using Ruby Gems. You can install it from command line:
```sh
  gem install wmap
```

### Specific Installation Problem with Nokogiri
Nokogiri is a native xml/html parser used by the project. It's fast and powerful. However, it comes with pitfall of installation problem around building native extension for your environment. Please refer to this page for trouble-shooting tip (http://www.nokogiri.org/tutorials/installing_nokogiri.html).

### Dependency
You need the Ruby 2.1.0 or above in order to use this program. In my test environment, I was able to set it up with <a href="https://rvm.io/">RVM</a>. Please refer to this page for more installation information: https://www.ruby-lang.org/en/documentation/installation/

In addition, the following Ruby GEM dependency are needed by different components of this software. The should be installed automatically:
      require "dnsruby"
      require "geoip"
      require "minitest/autorun"
      require "net/ping"
      require "netaddr"
      require "nokogiri"
      require "css_parser"
      require "openssl"
      require "open_uri_redirections"
      require "parallel"
      require "whois"
      require 'httpclient'


In case you want to install the above gems separately, use the command below:

      gem install dnsruby geoip minitest net-ping netaddr nokogiri css_parser open_uri_redirections openssl parallel whois httpclient

### Ruby-whois Gem Patches
This software depends on a patched version of Ruby gem ruby-whois (http://www.ruby-whois.org/) for the domain whois lookup feature. For better result, you could manually add the patches into your local whois gem installation directory as shown below:
```sh
      cp whois_patches/* [Your_ruby_whois_gem_path]/whois/lib/whois/record/parser/
```

Or you can directly download the branched whois gem from this repository - https://github.com/yangsec888/whois


### Before Using This Program
You need to define a scope for the program to run successful. The scope includes both your legitimate Internet domain, and your public
network block in the CIDR format.

To add your Internet domain into the scope, use the build-in shell command below:
```sh
     trust XYZ.COM
``
To add your public network block into the scope (note current support of IPv4 only):
```sh
     trust x.x.x.x/x
```

### Automatic Discovery and Tracking
```sh
    wmap <seed file | target host | target url | target IP or network cidr>
```
The above utility is intelligent enough to take argument as either a seed file, or a string such as a host, an IP, a network block, or a URL. The new discoveries will be automatically tracked in the data file 'lib/wmap/data/target_sites'.
  Note: seed file - mix of url, cidr and domain seed, one entry per line.
				url seed - known URL(s) for further discovery via the web crawler.
				cidr seed - list of known network blocks, for discovering web service via port scanning; it is also used to validate if the web service has a known IP (internal hosted).
				domain seed - validated internet domain to be used for DNS record brute-forcing; it is also used to validate the ownership of found web service.


### Dump Out Discovery Database
You can dump out the program output by using the build-in utility 'wdump' as shown below:
```sh
    wdump [output file name from you]
```
The above utility will dump out the discovery database into a single file as program output. Currently, the supported file format is Comma-separated Value (.csv) and Extensible Markup Language (.xml)


### More Usage Cases:
There are more examples under the 'demos' folder of this package. The examples show how to use the 'wmap' API to get your job done easily. Please check out the code - they should be easy and straightforward to be understood.


## More Document(s):
The software comes with the Ruby doc during your installation as shown above. For your convenience, the Ruby doc is also distributed with this software. You can navigate to the 'doc' folder of your local installation, and click the 'index.html' to open the start page in your favorite browser. You can also download the wmap-x.x.x.rdoc.zip documentation package alone from GitHub, unzip and open the doc/index.html in your browser.

If you need additional documentation / information other than this README file and the Ruby document package, please be patient - as I'm still working on it :)

## Bug Report or Feature Request?
Contact the author Sam Li directly at email 'yang.li@owasp.org'.


## Legal Disclaimer:
This software is provided strictly 'as-if' without any implied warranty. You're free to copy or modify the codes anyway you want - a reference back to this software will be appreciated. Please refer to the 'LICENSE.txt' file for more information.


CHANGELOG.md
##
# Wmap
#
- A pure Ruby library for Internet web application discovery and tracking.
-
- Copyright (c) 2012-2015 Yang Li <yang.li@owasp.org>


# Change-log


## Mile-stones

- November 2018: Pick up the maintenances.
- July 2015: Move the project under OWASP
- November 2014: Re-name from web_discovery to wmap, re-factor the code base to better scale up
	across the board
- May 2013: First release of web_discovery gem package 1.1.0
- December 2012: Alpha Release 0.1.0, with everything in one file for fast prototype purpose


## Backward Incompatibilities

- List of features that are backward incompatible:


## Beta Release 1.x

- 07/25/2015  Clean up for OWASP release.
-	04/14/2015	Fix bug in 'trust'/'distrust' utilities, so that they could properly handle CIDR as input format.
-	04/02/2015	Removing tracking of the 'un-reponsive' http site under 'Wmap::SiteTracker' class.
-	03/30/2015	Fix a bug in 'bulk_refresh' method under 'Wmap::SiteTracker' class.
-	03/24/2015	Implement 'wcheck' executable support to improve user experience.
-	03/23/2015	Implement 'wadds' executable support to improve user experience.  
-	02/19/2015	Bug fix on url_2_site method to handle rare cases containing "-" or "?" special chars.
-	02/18/2015	Implement the tcp port-scanner executable 'wscan'.
-	02/17/2015	A bug fix on the Wmap::Utils.normalize_url method, where the rare case of trailing dot after hostname could
-				  properly removed as well.
-	02/11/2015	A bug fix on the Wmap::SiteTracker.get_prim_uniq_sites method, where host resolved to multiple IPs could
-				  cause exception in the rare case.
-	02/10/2015	Implement the 'singleton' module in the Wmap::HostTracker::PrimaryHost class
-	02/05/2015	Implement singleton pattern on the Wmap::DomainTracker.instance::SubDomain, Wmap::SiteTracker::DeactivatedSite class.  
-	02/03/2015	Implement singleton pattern on the logger module, re-organize the log file structure and location.
-	01/30/2015	Implement the singleton pattern on the Wmap::SiteTracker, Wmap::DomainTracker.instance class.
-	01/27/2015	Separate the logger sub-module and implement the singleton pattern on the logger.
-	01/24/2015	Implement the 'singleton' module in the Wmap::HostTracker class, in order to avoid race condition
-				  under the parallel engine
-	01/08/2015	Implement the Ruby MiniTest Unit-test frame-work, with the first unit test file 'utils_test.rb'
-	12/15/2014	Replace the instance variables @known_hosts, @known_sites with class variables
-				 across all modules,i.e. enforcing the singleton pattern to avoid race condition under parallel execution.
-	11/17/2014	Re-factor the code Wmap::SiteTracker, optimize the 'dump' and 'dump_xml' methods.
-	11/14/2014	Change the package name from "WebDiscovery" version 1.5.3 to "Wmap" to simple reason, start with version 1.0  
-	11/13/2014	Add XML support as the program output format, i.e. 'save_uniq_sites_xml' method for WebDiscovery::SiteTracker class
-	11/01/2014	Add 'add' and 'delete' methods for WebDiscovery::CidrTracker class, in order to make it self-contained and user friendly.
-	10/31/2014	Add executables (trust, distrust etc..) under bin directory, so that the application is more user friendly
-	10/28/2014	Add parallel support in the HostTracker, SubDomain classes.
-	10/27/2014	Add 'prime', 'wdump' bin executables, to manually set the prime host, and to dump out unique sites
-				 in the site tracking data repository respectively.
-	10/21/2014	Re-implement bulk_refresh method under WebDiscovery::SiteTracker class,
-				 with parallel engine support.
-	09/29/2014	Add 'save' method to WebDiscovery::UrlCrawler class.
-	09/08/2014	Support parallel and add 'checks' method to WebDiscovery::UrlChecker class, in order to
-				 scale up.
-	08/04/2014	Add brute_all method for WebDiscovery::DnsBruter class.
-	07/21/2014	Add consistency checks for WebDiscovery::SiteTracker::Dump method.
-	07/08/2014	Add stop_hostname method in the class WebDiscovery::HostTracker; enhance the DnsBruter methods
-				 accordingly.
-	05/19/2014	Add site_ip_known? method in the class WebDiscovery::SiteTracker
-	03/31/2014	Add deduplicate procedures in the class WebDiscovery::HostTracker::PrimaryHost
-	02/10/2014	Add additional class WebDiscovery::DomainTracker::SubDomain to better manage sub-domains
-				 make changes in DomainTracker, HostTracker, and executable 'wd' accordingly
-	02/07/2014	Add additional class WebDiscovery::SiteTracker::DeactivatedSite and logic to record
-				 the decommissioned or no longer accessible site
-	01/21/2014	Introduce 'hostname_mutation' method into WebDiscovery::DnsBruter class
-	01/15/2014	Modify WebDiscovery::Utils::UrlMagic class and add URL 'normalize' method
-	01/14/2014	Modify WebDiscovery::UrlCrawler class and set hard stop limit of crawler time-out to
-				1 0 minutes per instance.
-   01/06/2014	Revamp WebDiscovery::UrlCrawler class, to make it more readable and scalable.
-   12/19/2013	Add new method 'SiteTracker.site_check', to pull out a record from the site store.
-   12/04/2013	Extend the usage of the 'parallel' to the 'SiteTracker.refresh_all' method, in order to
-				 scale up the completion time.
-   11/06/2013	Include GeoIPLite into the gem pack and modify the links under GeoIPTracker class.
-   10/29/2013	Fix a small bug in the 'HostTracker.bulk_add' method
-   10/21/2013	Change the data structure of domain_tracker class, to include open zone transfer
-                information.
-   10/09/2013	Add additional logic to handle Dnsruby::ServFail type error intelligently in order to
-                optimize the DnsBruter speed.
-   10/08/2013	Add additional logic to avoid repeating crawling the same link by different child processes.
-   10/04/2013	Add additional logic to profile the web server to improving the crawler speed
-   09/30/2013	Add additional logic to eliminate crawling duplicate sites for the multi-threaded crawlers.

demos/rpt_index.rb
#!/usr/bin/ruby
# Simple script to create an index.html page for the Dynamic MP summary report
# Usage: ./rpt_index.rb [list of report paths] 
################################################################################################
# Developed by: Yang Li, Version: 0.8, Last Modification: 09/28/2013
################################################################################################
require "resolv"

@verbose=true			# set debugging bit
@MAP=Hash.new		# Store site mapping matrix
@RPT_DB=Hash.new		# Store the final report matrix 
@INDEX_HTML="index.html"
@Map_file="Q1_map_key.csv"

def url_2_host (url)
# Extract Fully Qualified Domain Name (FQDN) from the url 
# For example: "https://login.yahoo.com/email/help" -> "login.yahoo.com"
	url = url.downcase.gsub(/(http:\/\/|https:\/\/)/, "")
	record1 = url.split(%r{\/})
	if record1[0].nil?
		puts "Error process url: #{url}" 
		return nil
	else
		record2 = record1[0].split(%r{\:})		
		return record2[0]
	end
end

def url_2_port (url)
	port=""
	if url=~/\:(\d+)/
		port=$1
	elsif url =~ /https/i
		port="443"
	else
		port="80"
	end
end

def url_2_site (url)
# Extract site in (host:port) format from the url 
# For example: "https://login.yahoo.com:8443/email/help" -> "login.yahoo.com:8443"
	port=url_2_port(url)
	host=url_2_host(url)
	site=host+":"+port
end

def file_2_list(f)
# loader for the reports
	list=Array.new
	file = File.open(f, "r")
	file.each_line do |line|
		list.push(line.chomp)
	end
	file.close
	return list
end

def load_map_from_file (map)
# load the site index (i.e. url => IP, Hosting, Netname, GeoIP, Res Code)
	begin
		f=File.open(map, 'r')
		f.each do |line|
			next if not line =~ /\,/
			puts "Loading: #{line.chomp}" if @verbose
			entry=line.chomp.split(%r{\,})
			site=url_2_site(entry[1])
			@MAP[site] = Hash.new if not @MAP.key?(site)
			@MAP[site]['index']=entry[0]
			@MAP[site]['url']=entry[1]
			@MAP[site]['ip'] = entry[2]
			@MAP[site]['hosting'] = entry[3]
			@MAP[site]['netname'] = entry[4]
			#@MAP[site]['rescode'] = entry[5]
			@MAP[site]['geoip'] = entry[5]
			@MAP[site]['server']=entry[6]
			@MAP[site]['status']=entry[7]
		end
		f.close
	rescue => ee			
		puts "Error: #{ee}" if @verbose
	end
end

def print_map
	puts "Summary of Site URI to SSE Lead Matrix:"
	@MAP.each do |key|
		name=@MAP[key]
		puts "#{key}, #{name}"
	end
	puts "End of the summary"
end

def file_path_2_host (path)
# extract host from the file path
	begin
		puts "Process file path: #{path}"
		file=path.split(%r{\/})[4]
		host=file.gsub(/^_/,'').gsub(/MP - /,'').gsub(/(http|https)\_+/,'').gsub(/\_*\.(pdf|xml)/,'').gsub(/\_\d+/,'')#.gsub(/\_+/,'')
		puts "Extracted host: #{host}"
		return host
	rescue => ee 
		puts "Exception on method #{__method__}: #{ee}"
	end
end

def file_path_2_rpt_name (path)
# extract report name from the file path
	file=path.split(%r{\/})[4]
	return file
end

def file_path_2_site(path)
# extract site in (host:port) format from the file path, as the unique identifier for the mapping process
	file=path.split(%r{\/})[4]
	puts "Extracted file: #{file}" if @verbose
	host=file_path_2_host(path)
	port=""
	site=host
	if file =~ /\_(\d+)\_\_\.(xml|pdf)/
		port=$1
		puts "Found uncommon web port from #{path}: #{$1}" if @verbose
	elsif file =~ /https/i
		port="443"
	else
		port="80"
	end
	site=host+":"+port
	puts "Extracted site: #{site}" if @verbose
	return site
end

def build_rpt_index (file)
# input is a list of report path. load it into the hash with tracking info, i.e. (site => link, IP)
	# load the report list 
	l = file_2_list(file)
	l.each do |x|	# build report db
		file=file_path_2_rpt_name(x)
		#host=file_path_2_host(x)
		site=file_path_2_site(x)
		link="<a href='#{x}'>#{file}</a>"
		@RPT_DB[site]=Hash.new unless @RPT_DB.key?(site)
		@RPT_DB[site]['link']=link
		@RPT_DB[site]['file']=file
		@RPT_DB[site]['hosting']=@MAP[site]['hosting'] if @MAP.key?(site)
		@RPT_DB[site]['ip']=@MAP[site]['ip'] if @MAP.key?(site)
		@RPT_DB[site]['geoip']=@MAP[site]['geip'] if @MAP.key?(site)
		@RPT_DB[site]['hosting']=@MAP[site]['hosting'] if @MAP.key?(site)
		#@RPT_DB[site]['rescode']=@MAP[site]['rescode'] if @MAP.key?(site)
		@RPT_DB[site]['netname']=@MAP[site]['netname'] if @MAP.key?(site)
		@RPT_DB[site]['url']=@MAP[site]['url'] if @MAP.key?(site)
		@RPT_DB[site]['server']=@MAP[site]['server'] if @MAP.key?(site)
		@RPT_DB[site]['status']=@MAP[site]['status'] if @MAP.key?(site)
		@RPT_DB[site]['index']=@MAP[site]['index'] if @MAP.key?(site)
		puts "build_rpt_index: #{site}, #{link}," if @verbose
	end
end

def print_rpt_index
	puts "Summary of Report Matrix:"
	@RPT_DB.each do |key,value|
		next if key.nil?
		link=@RPT_DB[key]['link']
		netname=@RPT_DB[key]['netname']
		puts "#{key}, #{link}, #{netname}"
	end
	puts "End of the summary"
end
################################################################################################
# Main 
################################################################################################
# load the inventory list into the application matrix
# Typical map entry: 
# "http://chconnect.barclays.com/,141.228.106.100,int_hosted,BARCLAYS-CIBWM,GBR H9 London,302 "
abort "Error: incorrect program argument. \nUsage: ./rpt_index.rb [list of report paths]" unless ARGV.size>0
load_map_from_file (@Map_file) 
print_map if @verbose
build_rpt_index (ARGV[0])
print_rpt_index if @verbose
# write the HTML index page
open(@INDEX_HTML, 'w') do |f|
	f.puts "<html>\n<title>Barclays Perimeter Application Scan Report Index Page</title>\n<body>"
	f.puts "<table border='1'>"
	f.puts "<tr><td>Index</td><td>Test Site</td><td>Report Link</td><td>Primary IP</td><td>Hosting Status</td><td>Netname</td><td>GeoIP Location</td><td>Server</td><td>Status Code</td></tr>"
	f.puts "<b>Barclays 2015 Q1 Perimeter Application Scan Report Index - Barclaycard US</b><p>"
	@MAP.each do |key,values|
		index=@MAP[key]['index']
		url=@MAP[key]['url']
		pip=@MAP[key]['ip']
		hosting=@MAP[key]['hosting']
		netname=@MAP[key]['netname']
		geoip=@MAP[key]['geoip']
		server=@MAP[key]['server']
		status=@MAP[key]['status']
		#rescode=@MAP[key]['rescode']
		site=key
		if @RPT_DB.key?(site)			
			link=@RPT_DB[site]['link'] if @RPT_DB.key?(site)
			f.puts "<tr>"
			f.puts "<td>#{index}</td><td>#{url}</td><td>#{link}</td><td>#{pip}</td><td>#{hosting}</td><td>#{netname}</td><td>#{geoip}</td><td>#{server}</td><td>#{status}</td>"
			f.puts "</tr>"
		else
			f.puts "<tr>"
			f.puts "<td>#{index}</td><td>#{url}</td><td>TBD</td><td></td><td></td><td></td><td></td><td></td>"
			f.puts "</tr>"
		end
	end
	f.puts "</table><p>"
	f.puts "<i><b>Further Info & Feedback</b>: Please contact the Barclays Application Security team at <a href='mailto:GRBApplicationSecurity@barclayscorp.com'>GRBApplicationSecurity@barclayscorp.com</a> for further information and / or any feedback. <p>"
	f.puts "<b>Barclays Internal Only</b>: Please restrict the report data from the public distribution outside of the Barclays.</i> "
	f.puts "</body></html>"
end
